presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -367,6 +376,24 @@ public void enforceMemoryLimits()
     }
 
     /**
+     * Enforce timeout at the query level
+     */
+    public void enforceQueryMaxRuntimeLimits()
+    {
+        Iterable<QueryExecution> runningQueries = queries.values().stream()
+                .filter(query -> query.getQueryInfo().getState() == RUNNING)
+                .collect(toImmutableList());
+
+        for (QueryExecution runningQuery : runningQueries) {
+            Duration queryMaxRuntime = SystemSessionProperties.getQueryMaxRuntime(runningQuery.getSession());
```

Comment:
1. Checking `isDone` can be considered a optimization. But it is not a guarantee. Status could change after you call `isDone`, but before you call `fail`. Adding a comment can solve the confusion. But I agree with @yaoxin226 that it's better if the code is self explanatory. I don't have a strong opinion on this issue.
2. The overhead is far from non-existent. But it's also not prohibitively high. That's why I raised the question to @dain above.

We will discuss again and give you an answer.